### PR TITLE
Tested changes for efficient SMC grid coupling

### PR DESCRIPTION
**Pull request by ukmo-juan-castillo**
Modifications to tackle issue 250:

https://github.com/NOAA-EMC/WW3/issues/250

This branch was successfully tested using the Met Office amm15 and indsl2 configurations, and I made sure that the OASIS grid file contained the same information as the operational version of the code, although with a different format. Jian-Guo provided the SMC grid that was added in the regtests, and he successfully run a propagation test.

The remap files from the toy model to the wave smc grid have to be added to data_regtests, as for some reason they could not be generated on the fly by OASIS and I had to generate them offline - I did not have this problem with the amm15 and indsl2 grids, though. I will attach them as separate files.

I will later provide a new commit with an update of the files ww3_grid_OASACM6.nml and ww3_shel_OASACM6.nml so that they are consistent with their *.inp versions.

### CODE REVIEW

**Comment by ukmo-juan-castillo** [comment]
These are the weight files to run the new regtest.

[weights.zip](https://github.com/ukmo-waves/WW3/files/5347030/weights.zip)

**Comment by mickaelaccensi** [comment]
Hi Juan,
is it still a draft PR or PR that you are waiting for a review now ? If so, could you add the results of the regtests comparison ?
otherwise you can set the PR as draft and once you are ready for review process, set it again to PR.

**Comment by ukmo-juan-castillo** [comment]
Attaching the result of the coupling regtests. There are small rounding differences in the OASACM and OASOCM tests, which I assume they are due to changing the order of some lines in w3oacpmd.f90 .
[matrixCompSummary.out-ww3_tp2.14.txt](https://github.com/ukmo-waves/WW3/files/5395628/matrixCompSummary.out-ww3_tp2.14.txt)

**Comment by ukmo-juan-castillo** [comment]
In fact the differences look like they are caused by different OASIS restart files, although supposedly I am using the same.

[regtests.zip](https://github.com/ukmo-waves/WW3/files/5395638/regtests.zip)

**Comment by mickaelaccensi** [comment]
ok I supposed the differences are due to the bug NOAA-EMC#259
could you redo the regtests once this bug is fixed to be sure ?

**Comment by ukmo-juan-castillo** [comment]
> ok I supposed the differences are due to the bug [NOAA-EMC#259](https://github.com/NOAA-EMC/WW3/issues/259)
> could you redo the regtests once this bug is fixed to be sure ?

Sure, I will do once the bug fix is merged to the staging branch, and I merge this branch to pick the bug fixes.

**Comment by ukmo-juan-castillo** [comment]
After updating the branch to merge the latest coupling bug fix, the restests show no differences in the original coupling tests - although there are some differences in the new OASACM2-5.
[regtests.tar.gz](https://github.com/ukmo-waves/WW3/files/5423434/regtests.tar.gz)

**Review by mickaelaccensi — APPROVED** [review]

File: model/ftn/w3oacpmd.ftn
Comment (ukmo-ccbunney) [inline model/ftn/w3oacpmd.ftn]: Only a small thing:
I think this comment is a bit misleading, the DLON and DLAT values are the smallest grid cell increments (i.e. the tier 1 cells). The number of tiers could (theoretically!) be set to any value, although 2, 3 or 4 are the likely to be the most common.

**Review by ukmo-ccbunney — APPROVED** [review]
These changes look good to me @ukmo-juan-castillo other than the one comment I have made regarding the DLAT/DLON comment.

**Comment by ukmo-juan-castillo** [comment]
> Modifications to tackle issue 250:
> 
> [NOAA-EMC#250](https://github.com/NOAA-EMC/WW3/issues/250)
> 
> This branch was successfully tested using the Met Office amm15 and indsl2 configurations, and I made sure that the OASIS grid file contained the same information as the operational version of the code, although with a different format. Jian-Guo provided the SMC grid that was added in the regtests, and he successfully run a propagation test.
> 
> The remap files from the toy model to the wave smc grid have to be added to data_regtests, as for some reason they could not be generated on the fly by OASIS and I had to generate them offline - I did not have this problem with the amm15 and indsl2 grids, though. I will attach them as separate files.
> 
> I will later provide a new commit with an update of the files ww3_grid_OASACM6.nml and ww3_shel_OASACM6.nml so that they are consistent with their *.inp versions.

The problem with the weight files has been traced to the grid file that OASIS generates. The grid file is perfectly fine, but because the domain used for the regtest is quite small and the grid dimensions do not contain a 'units' attribute, the ECMF program used to generate the weight files interprets the lat and lon units as radians, when they are in fact degrees. This should not be a problem when running with larger domains, or in the worst case, one can add the unites to the grid file generated by OASIS by hand.

**Pull request merged**
